core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala

```diff
@@ -17,33 +17,84 @@
 
 package org.apache.spark
 
-import akka.actor.Actor
+import scala.concurrent.duration._
+import scala.collection.mutable
+
+import akka.actor.{Actor, Cancellable}
+
 import org.apache.spark.executor.TaskMetrics
 import org.apache.spark.storage.BlockManagerId
-import org.apache.spark.scheduler.TaskScheduler
+import org.apache.spark.scheduler.{SlaveLost, TaskScheduler}
 import org.apache.spark.util.ActorLogReceive
 
 /**
  * A heartbeat from executors to the driver. This is a shared message used by several internal
- * components to convey liveness or execution information for in-progress tasks.
+ * components to convey liveness or execution information for in-progress tasks. It will also 
+ * expire the hosts that have not heartbeated for more than spark.driver.executorTimeoutMs.
  */
 private[spark] case class Heartbeat(
     executorId: String,
     taskMetrics: Array[(Long, TaskMetrics)], // taskId -> TaskMetrics
     blockManagerId: BlockManagerId)
 
+private[spark] case object ExpireDeadHosts 
+    
 private[spark] case class HeartbeatResponse(reregisterBlockManager: Boolean)
 
 /**
  * Lives in the driver to receive heartbeats from executors..
  */
-private[spark] class HeartbeatReceiver(scheduler: TaskScheduler)
+private[spark] class HeartbeatReceiver(sc: SparkContext, scheduler: TaskScheduler)
   extends Actor with ActorLogReceive with Logging {
 
+  val executorLastSeen = new mutable.HashMap[String, Long]
+  
+  val executorTimeout = sc.conf.getLong("spark.driver.executorTimeoutMs", 
+    sc.conf.getLong("spark.storage.blockManagerSlaveTimeoutMs", 120 * 1000))
+  
+  val checkTimeoutInterval = sc.conf.getLong("spark.driver.executorTimeoutIntervalMs",
```

Comment:
> Everywhere else spark.network.timeout is used there might already be garbage collection

You are right.

However, in all the other places that `spark.network.timeout` is used, there's another more specific property that can override it, e.g. `spark.shuffle.io.connectionTimeout`, `spark.core.connection.ack.wait.timeout`, `spark.akka.timeout`.  I think it would be useful to have the same thing here.  (If that's what you were saying in the first place here, sorry for the noise).